File: leetcode/0001-1000/0238-product-of-array-except-self/ProductOfArrayExceptSelf.java
import java.util.Arrays;

public class ProductOfArrayExceptSelf {
  public static void main(String[] args) {
    int[] numbers = new int[] {
        1, 2, 3, 4
    };

    System.out.println(Arrays.toString(new Solution().productExceptSelf(numbers)));
  }
}

class Solution {
  public int[] productExceptSelf(int[] numbers) {
    int totalZeros = 0;
    int zeroIndex = 0;
    int product = 1;

    for (int i = 0; i < numbers.length; i++) {
      if (numbers[i] == 0) {
        totalZeros++;
        zeroIndex = i;
        continue;
      }

      product *= numbers[i];
    }

    if (totalZeros > 0) {
      Arrays.fill(numbers, 0);

      if (totalZeros == 1) {
        numbers[zeroIndex] = product;
      }
    } else {
      for (int i = 0; i < numbers.length; i++) {
        numbers[i] = product / numbers[i];
      }
    }

    return numbers;
  }
}

File: leetcode/0001-1000/0201-0300/0231-0240/0238-product-of-array-except-self/ProductOfArrayExceptSelf.ts
function productExceptSelf(numbers: number[]): number[] {
  const { totalZeros, product, lastZeroIndex } = numbers.reduce(
    (previousAccumulated, currentValue, currentIndex) => {
      const { totalZeros, product, lastZeroIndex } = previousAccumulated;

      if (currentValue === 0) {
        return {
          totalZeros: totalZeros + 1,
          product,
          lastZeroIndex: currentIndex
        };
      }

      return {
        totalZeros,
        product: product * currentValue,
        lastZeroIndex
      };
    },

    { totalZeros: 0, product: 1, lastZeroIndex: -1 }
  );

  if (totalZeros >= 2) {
    return numbers.fill(0);
  }

  if (totalZeros === 1) {
    numbers.fill(0);

    numbers[lastZeroIndex] = product;

    return numbers;
  }

  numbers.forEach(
    (value, index) => {
      numbers[index] = product / value;
    }
  )

  return numbers;
};

console.log(productExceptSelf([1, 2, 3, 4, 5]));
console.log(productExceptSelf([1, 0, 3, 4, 5]));
console.log(productExceptSelf([1, 0, 3, 0, 5]));
